test/e2e/xfn_test.go

```diff
@@ -0,0 +1,318 @@
+				Namespace: ns,
+			},
+			Type: corev1.SecretTypeTLS,
+			StringData: map[string]string{
+				"tls.crt": caPem,
+				"tls.key": keyPem,
+			},
+		}
+		client := config.Client().Resources()
+		if err := client.Create(ctx, secret); err != nil {
+			t.Fatalf("Cannot create secret %s: %v", secret.Name, err)
+		}
+		configMap := &corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "reg-ca",
+				Namespace: namespace,
+			},
+			Data: map[string]string{
+				"domain.crt": caPem,
+			},
+		}
+		if err := client.Create(ctx, configMap); err != nil {
+			t.Fatalf("Cannot create config %s: %v", configMap.Name, err)
+		}
+		return ctx
+	}
+}
+
+// CopyImagesToRegistry copies fn images to private registry
+func CopyImagesToRegistry() features.Func {
+	return func(ctx context.Context, t *testing.T, config *envconf.Config) context.Context {
+		nodes := &corev1.NodeList{}
+		if err := config.Client().Resources().List(ctx, nodes); err != nil {
+			t.Fatal("cannot list nodes", err)
+		}
+		if len(nodes.Items) == 0 {
+			t.Fatalf("no nodes in the cluster")
+		}
+		i := slices.IndexFunc(nodes.Items[0].Status.Addresses, func(a corev1.NodeAddress) bool {
+			return a.Type == corev1.NodeInternalIP
+		})
+		if i == -1 {
+			t.Fatalf("no nodes with private address")
+		}
+		addr := nodes.Items[0].Status.Addresses[i].Address
+		p := gexe.RunProc(fmt.Sprintf("skopeo copy docker-daemon:crossplane-e2e/fn-labelizer:latest docker://%s:32000/fn-labelizer:latest --dest-tls-verify=false", addr)).Wait()
+		out, _ := io.ReadAll(p.Out())
+		t.Logf("skopeo stdout: %s", string(out))
+		if p.ExitCode() != 0 {
+			t.Fatalf("copying image to registry not successful, exit code %v std out %v std err %v", p.ExitCode(), string(out), p.Err())
+		}
+		return ctx
+	}
+}
+
+// InstallDockerRegistry with custom TLS
+func InstallDockerRegistry() features.Func {
+	return funcs.AsFeaturesFunc(funcs.HelmInstall(
+		helm.WithName("private"),
+		helm.WithNamespace("reg"),
+		helm.WithWait(),
+		helm.WithChart("https://helm.twun.io/docker-registry-2.2.1.tgz"),
+		helm.WithArgs(
+			"--set service.type=NodePort",
+			"--set service.nodePort=32000",
+			"--set tlsSecretName=reg-cert",
+		),
+	))
+}
+
+// CrossplaneDeployedWithFunctionsEnabled asserts that crossplane deployment with composition functions is enabled
+func CrossplaneDeployedWithFunctionsEnabled() features.Func {
+	return funcs.AllOf(
+		funcs.AsFeaturesFunc(funcs.HelmUpgrade(
+			HelmOptions(
+				helm.WithArgs(
+					"--set args={--debug,--enable-composition-functions}",
+					"--set xfn.enabled=true",
+					"--set xfn.args={--debug}",
+					"--set registryCaBundleConfig.name=reg-ca",
+					"--set registryCaBundleConfig.key=domain.crt",
+				))...)),
+		funcs.ReadyToTestWithin(1*time.Minute, namespace),
+	)
+}
+
+// ProvideNopDeployed assets that provider-nop is deployed and healthy
+func ProvideNopDeployed() features.Func {
+	manifests := "test/e2e/manifests/apiextensions/composition/minimal/prerequisites"
```

Comment:
see comment above, let's avoid relying on other tests' prerequisites as that could lead to hard to debug issues.